azure/src/main/java/org/apache/iceberg/azure/blob/AzureBlobInputStream.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.azure.blob;
+
+import com.azure.storage.blob.BlobClient;
+import com.azure.storage.blob.models.BlobRange;
+import com.azure.storage.blob.options.BlobInputStreamOptions;
+import com.azure.storage.blob.specialized.BlobInputStream;
+import java.io.IOException;
+import java.util.Arrays;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.io.SeekableInputStream;
+import org.apache.iceberg.relocated.com.google.common.base.Joiner;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AzureBlobInputStream extends SeekableInputStream {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AzureBlobInputStream.class);
+  private static final Joiner TRACE_JOINER = Joiner.on("\n\t");
+
+  private final StackTraceElement[] createStack;
+  private final AzureURI azureURI;
+  private final AzureProperties azureProperties;
+  private final BlobClient blobClient;
+
+  private long pos = 0L;
+  private boolean closed = false;
+  private BlobInputStream stream;
+
+  public AzureBlobInputStream(AzureURI azureURI, AzureProperties azureProperties, BlobClient blobClient) {
+    this.createStack = Thread.currentThread().getStackTrace();
+    this.azureURI = azureURI;
+    this.azureProperties = azureProperties;
+    this.blobClient = blobClient;
+    openStream(/* offset = */ 0);
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "Stream already closed.");
+    Preconditions.checkArgument(newPos >= 0, "New position cannot be negative: %s.", newPos);
+
+    if (newPos == pos) {
+      // Already at the specified position.
+      return;
+    }
+
+    if (newPos > pos) {
+      // Seeking forward.
+      final long bytesToSkip = newPos - pos;
+      // BlobInputStream#skip only repositions the internal pointers,
+      // the actual bytes are skipped when BlobInputStream#read is invoked.
+      final long bytesSkipped = stream.skip(bytesToSkip);
```

Comment:
Does this read through or just change the next request?

Reply:
It does not read through, it repositions internal pointers to change the next read request

https://github.com/Azure/azure-sdk-for-java/blob/627f10710712d00eea89dfd892587f18da2a456a/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StorageInputStream.java#L419

https://github.com/Azure/azure-sdk-for-java/blob/627f10710712d00eea89dfd892587f18da2a456a/sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StorageInputStream.java#L379